README.md
# photoEditor

Фоторедактор, реализованный на Python + PyQt5 + Pillow

## Installation

`git clone https://gitlab.com/shmd001/photoeditor.git`  
`cd photoeditor`  
`pip install -r requirements.txt`  
`python main.py`  

## [CHANGELOG](./CHANGELOG.md)

## Notes:

- Регистрация пользователей осуществляется путем ввода нового логина, ненаходящегося в базе данных
- Фильтр Сепия работает дольше остальных из-за особенностей его разработки
- Сохранение файла в нужное место осуществляется путем нажатия соответсвующей кнопки в окне редактирования


## Developers Team:
    1. Maintainer - 2020-5-24-sho
    2. Developer  - 2020-5-21-tak
    3. Developer  - 2020-5-01-abr
    4. Developer  - 2020-5-23-shi

## Цели:
    1. Реализовать обработку исключений, добавить вывод ошибок в виде QErrorMessage
    2. Напиать тесты, протестировать

## Важная информация:
В связи с переходом на редактирование изображения с помощью библиотеки Pillow, меняю задачи  
Постарайтесь сделать как можно быстрее, объем работ будет увеличен

### Задачи для maintainer:
* [x] Добавить README.md с текущими целями
* [x] Добавить requirements.txt, со списком всех библиотек, используемых в проекте
* [x] Пересобрать главный класс фоторедактора, сделать формирование дизайна функционально
* [x] Реализовать ```createMainLabel()```, рисующую главный label
* [x] Убрать ненужный функционал при открытии главного окна приложения
* [x] Написать стили для текущих компонентов, удобно расположить компоненты на окне
* [x] Связать функции наложения фильтров с нажатием на соответсвующие кнопки
* [x] Реализовать хранение файлов и убрать неактуальные свойства класса Image
* [x] Спроектировать и создать БД, реализовать класс для работы с ней
* [x] Связать все окна между собой (подробное описание в issue)

### Задачи для developer-2:
* [x] Реализовать запуск окна, согласно дизайну 
* [x] Реализовать ```createEditingBar```, рисующую панель редактирования
* [x] Привести названия переменных и функций к единому стилю 
* [x] Реалзовать открытие файла
* [x] Реалзовать функцию ```saveImage```, сохраняющую изображение
* [x] Реализовать наложение фильтра Сепия
* [x] Реализовать с помощью библиотеки Pillow наложение фильтров
* [x] Покрыть код комментариями для упрощения читабельности
* [x] Спроектировать и создать окно выбора изображения из списка изображений пользователя
* [x] Реализовать хранение паролей в зашифрованном виде

### Задачи для developer-3:
* [x] Сфорировать **design.py** согласно **design.ui**
* [x] Реализовать ```createMenu```, рисующую меню
* [x] Покрыть код комментариями  
* [x] Добавить иконки к элементам меню
* [x] Реализовать функцию ```updateActions``` (Описание укажу в issue)
* [x] Переделать все названия QActions на русский язык
* [x] Реализовать наложение фильтра Негатив
* [x] Реализовать отражение и поворот изображения с помощью Pillow
* [x] Добавить новые стили
* [x] Спроектировать и создать окно авторизации/регистрации пользователей
* [x] Покрыть код комментариями, задокументировать все классы и функции

### Задачи для developer-4:
* [x] Задизайнить окно фоторедактора в QT Designer, на выходе **design.ui**
* [x] Реализовать ```createToolBar```, рисующую панель инструментов
* [x] Разбить ```main.py``` на несколько файлов
* [x] Реализовать закрытие приложения нажатием на кнопку
* [x] Реализовать функцию ```revertToOriginal``` (Описание укажу в issue)
* [x] Реализовать наложение фильтра ЧБ
* [x] Пофиксить все иконки, отказаться от зума фотографии на экране и обрезания фото
* [x] Реализовать сброс всех кнопок при отмене изменений или загрузке новой фотографии
* [x] Рефакторинг кода, исправить баги
* [x] Повысить качество кода с помощью pylint, отформатировать с помощью black

### Задачи на перспективу:
* [x] Сформировать *about* и реализовать в виде MessageBox принажатии на кнопку в меню
* [ ] Создать для каждого исключения его обработку и вывод в `QErrorMessage`


CHANGELOG.md
# 2.0.0 (2021-12-22)

### Bug fixes (2 changes)

- Добавлена обработка ошибки, возникающая при невыбранном изображении
- Удалено действие Open

### Features (12 changes)

- Реализовано шифрование обозначенным способом
- Реализовано создание директории при регистрации
- Реализован возврат к списку изображений
- Реализовано открытие окна фоторедактора после выбора изображения
- Реализовано добавление в список нового изображения
- Реализовано добавление нового изображения в базу
- Добавлена возможность выбора нового изображения
- Добавлена возможность выхода из учетной записи
- Реализована регистрация нового пользователя
- Реализован запуск окна авторизации
- Действие Exit изменено на Back
- Добавлена база данных


# 1.0.0 (2021-12-17)


### Bug fixes (6 changes)

- Исправлен баг при неоткрытом изображении
- Исправлен баг: добавлена инициализация значений слайдеров в конструкторе
- Отказ от функционала увеличения изображения
- Отказ от функционала обрезки изображения
- Исправлен недостаток с отменой редактирования
- Реализована обработка ошибок при сохранении файла

### Features (8 changes)
- Реализовано наложение фильтров: ЧБ, Негатив, Сепия
- Реализован поворот изображения на 90 градусов
- Реализован флип изображения
- Реализовано изменение яркости и контраста изображения
- Реализовано копирование и хранение редактируемых файлов
- Реализовано сохранение отредактированного файла
- Добавлены стили
- Добавлены иконки


design.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'design.ui'
#
# Created by: PyQt5 UI code generator 5.15.6
#
# WARNING: Any manual changes made to this file will be lost when pyuic5 is
# run again.  Do not edit this file unless you know what you are doing.


from PyQt5 import QtCore, QtGui, QtWidgets


class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        MainWindow.setObjectName("MainWindow")
        MainWindow.resize(837, 518)
        self.centralwidget = QtWidgets.QWidget(MainWindow)
        self.centralwidget.setObjectName("centralwidget")
        self.gridWidget = QtWidgets.QWidget(self.centralwidget)
        self.gridWidget.setGeometry(QtCore.QRect(10, 10, 640, 480))
        self.gridWidget.setMaximumSize(QtCore.QSize(640, 480))
        self.gridWidget.setObjectName("gridWidget")
        self.gridLayout = QtWidgets.QGridLayout(self.gridWidget)
        self.gridLayout.setContentsMargins(0, 0, 0, 0)
        self.gridLayout.setObjectName("gridLayout")
        self.label_3 = QtWidgets.QLabel(self.gridWidget)
        self.label_3.setText("")
        self.label_3.setPixmap(QtGui.QPixmap("../Pictures/ibooks.ico"))
        self.label_3.setAlignment(QtCore.Qt.AlignCenter)
        self.label_3.setObjectName("label_3")
        self.gridLayout.addWidget(self.label_3, 0, 0, 1, 1)
        MainWindow.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(MainWindow)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 837, 21))
        self.menubar.setAutoFillBackground(False)
        self.menubar.setDefaultUp(False)
        self.menubar.setNativeMenuBar(True)
        self.menubar.setObjectName("menubar")
        self.menuPhotoEditor = QtWidgets.QMenu(self.menubar)
        self.menuPhotoEditor.setObjectName("menuPhotoEditor")
        self.menuFile = QtWidgets.QMenu(self.menubar)
        self.menuFile.setObjectName("menuFile")
        self.menuEdit = QtWidgets.QMenu(self.menubar)
        self.menuEdit.setObjectName("menuEdit")
        self.menuTools = QtWidgets.QMenu(self.menubar)
        self.menuTools.setObjectName("menuTools")
        MainWindow.setMenuBar(self.menubar)
        self.dockWidget_2 = QtWidgets.QDockWidget(MainWindow)
        self.dockWidget_2.setMinimumSize(QtCore.QSize(139, 38))
        self.dockWidget_2.setObjectName("dockWidget_2")
        self.dockWidgetContents_2 = QtWidgets.QWidget()
        self.dockWidgetContents_2.setObjectName("dockWidgetContents_2")
        self.verticalLayoutWidget = QtWidgets.QWidget(self.dockWidgetContents_2)
        self.verticalLayoutWidget.setGeometry(QtCore.QRect(20, 0, 101, 251))
        self.verticalLayoutWidget.setObjectName("verticalLayoutWidget")
        self.verticalLayout = QtWidgets.QVBoxLayout(self.verticalLayoutWidget)
        self.verticalLayout.setContentsMargins(0, 0, 0, 0)
        self.verticalLayout.setObjectName("verticalLayout")
        self.checkBox = QtWidgets.QCheckBox(self.verticalLayoutWidget)
        self.checkBox.setObjectName("checkBox")
        self.verticalLayout.addWidget(self.checkBox)
        self.checkBox_3 = QtWidgets.QCheckBox(self.verticalLayoutWidget)
        self.checkBox_3.setObjectName("checkBox_3")
        self.verticalLayout.addWidget(self.checkBox_3)
        self.checkBox_2 = QtWidgets.QCheckBox(self.verticalLayoutWidget)
        self.checkBox_2.setObjectName("checkBox_2")
        self.verticalLayout.addWidget(self.checkBox_2)
        self.label = QtWidgets.QLabel(self.verticalLayoutWidget)
        self.label.setEnabled(True)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Ignored, QtWidgets.QSizePolicy.Ignored)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.label.sizePolicy().hasHeightForWidth())
        self.label.setSizePolicy(sizePolicy)
        self.label.setAlignment(QtCore.Qt.AlignCenter)
        self.label.setObjectName("label")
        self.verticalLayout.addWidget(self.label)
        self.sliderigntness = QtWidgets.QSlider(self.verticalLayoutWidget)
        self.sliderigntness.setMaximum(99)
        self.sliderigntness.setSingleStep(10)
        self.sliderigntness.setProperty("value", 50)
        self.sliderigntness.setOrientation(QtCore.Qt.Horizontal)
        self.sliderigntness.setTickPosition(QtWidgets.QSlider.TicksAbove)
        self.sliderigntness.setTickInterval(10)
        self.sliderigntness.setObjectName("sliderigntness")
        self.verticalLayout.addWidget(self.sliderigntness)
        self.label_2 = QtWidgets.QLabel(self.verticalLayoutWidget)
        self.label_2.setEnabled(True)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Ignored, QtWidgets.QSizePolicy.Ignored)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.label_2.sizePolicy().hasHeightForWidth())
        self.label_2.setSizePolicy(sizePolicy)
        self.label_2.setStyleSheet("")
        self.label_2.setAlignment(QtCore.Qt.AlignCenter)
        self.label_2.setObjectName("label_2")
        self.verticalLayout.addWidget(self.label_2)
        self.slideContrast = QtWidgets.QSlider(self.verticalLayoutWidget)
        self.slideContrast.setMaximum(99)
        self.slideContrast.setSingleStep(10)
        self.slideContrast.setProperty("value", 50)
        self.slideContrast.setOrientation(QtCore.Qt.Horizontal)
        self.slideContrast.setTickPosition(QtWidgets.QSlider.TicksAbove)
        self.slideContrast.setTickInterval(10)
        self.slideContrast.setObjectName("slideContrast")
        self.verticalLayout.addWidget(self.slideContrast)
        self.dockWidget_2.setWidget(self.dockWidgetContents_2)
        MainWindow.addDockWidget(QtCore.Qt.DockWidgetArea(1), self.dockWidget_2)
        self.actionAbout = QtWidgets.QAction(MainWindow)
        icon = QtGui.QIcon()
        icon.addPixmap(QtGui.QPixmap("../../../python/photoeditor/resources/ibooks.ico"), QtGui.QIcon.Normal, QtGui.QIcon.Off)
        self.actionAbout.setIcon(icon)
        self.actionAbout.setObjectName("actionAbout")
        self.actionQuit = QtWidgets.QAction(MainWindow)
        icon1 = QtGui.QIcon()
        icon1.addPixmap(QtGui.QPixmap("../../../python/photoeditor/resources/close.png"), QtGui.QIcon.Normal, QtGui.QIcon.Off)
        self.actionQuit.setIcon(icon1)
        self.actionQuit.setObjectName("actionQuit")
        self.actionOpen = QtWidgets.QAction(MainWindow)
        self.actionOpen.setObjectName("actionOpen")
        self.actionSave = QtWidgets.QAction(MainWindow)
        self.actionSave.setObjectName("actionSave")
        self.actionRevert_to_original = QtWidgets.QAction(MainWindow)
        self.actionRevert_to_original.setObjectName("actionRevert_to_original")
        self.actionRotate_90 = QtWidgets.QAction(MainWindow)
        self.actionRotate_90.setObjectName("actionRotate_90")
        self.actionRotate_91 = QtWidgets.QAction(MainWindow)
        self.actionRotate_91.setObjectName("actionRotate_91")
        self.actionFlip_horizontal = QtWidgets.QAction(MainWindow)
        self.actionFlip_horizontal.setObjectName("actionFlip_horizontal")
        self.actionFlip_vartical = QtWidgets.QAction(MainWindow)
        self.actionFlip_vartical.setObjectName("actionFlip_vartical")
        self.menuPhotoEditor.addAction(self.actionAbout)
        self.menuPhotoEditor.addSeparator()
        self.menuPhotoEditor.addAction(self.actionQuit)
        self.menuFile.addAction(self.actionOpen)
        self.menuFile.addAction(self.actionSave)
        self.menuEdit.addAction(self.actionRevert_to_original)
        self.menuTools.addAction(self.actionRotate_90)
        self.menuTools.addAction(self.actionRotate_91)
        self.menuTools.addSeparator()
        self.menuTools.addAction(self.actionFlip_horizontal)
        self.menuTools.addAction(self.actionFlip_vartical)
        self.menubar.addAction(self.menuPhotoEditor.menuAction())
        self.menubar.addAction(self.menuFile.menuAction())
        self.menubar.addAction(self.menuEdit.menuAction())
        self.menubar.addAction(self.menuTools.menuAction())

        self.retranslateUi(MainWindow)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

    def retranslateUi(self, MainWindow):
        _translate = QtCore.QCoreApplication.translate
        MainWindow.setWindowTitle(_translate("MainWindow", "MainWindow"))
        self.menuPhotoEditor.setTitle(_translate("MainWindow", "Photo Editor"))
        self.menuFile.setTitle(_translate("MainWindow", "File"))
        self.menuEdit.setTitle(_translate("MainWindow", "Edit"))
        self.menuTools.setTitle(_translate("MainWindow", "Tools"))
        self.checkBox.setText(_translate("MainWindow", "Сепия"))
        self.checkBox_3.setText(_translate("MainWindow", "Черно-белый"))
        self.checkBox_2.setText(_translate("MainWindow", "Негатив"))
        self.label.setText(_translate("MainWindow", "Яркость"))
        self.label_2.setText(_translate("MainWindow", "Контраст"))
        self.actionAbout.setText(_translate("MainWindow", "About"))
        self.actionQuit.setText(_translate("MainWindow", "Quit"))
        self.actionQuit.setShortcut(_translate("MainWindow", "Ctrl+Q"))
        self.actionOpen.setText(_translate("MainWindow", "Open"))
        self.actionSave.setText(_translate("MainWindow", "Save"))
        self.actionRevert_to_original.setText(_translate("MainWindow", "Revert to original"))
        self.actionRotate_90.setText(_translate("MainWindow", "Rotate 90 ->"))
        self.actionRotate_91.setText(_translate("MainWindow", "Rotate 90 <-"))
        self.actionFlip_horizontal.setText(_translate("MainWindow", "Flip horizontal"))
        self.actionFlip_vartical.setText(_translate("MainWindow", "Flip vartical"))

if __name__ == "__main__":
    import sys
    app = QtWidgets.QApplication(sys.argv)
    MainWindow = QtWidgets.QMainWindow()
    ui = Ui_MainWindow()
    ui.setupUi(MainWindow)
    MainWindow.show()
    sys.exit(app.exec_())
